hazelcast-sql/src/test/java/com/hazelcast/sql/impl/expression/CastEndToEndTest.java

```diff
@@ -0,0 +1,369 @@
+                "Cast function cannot convert value of type DECIMAL(38, 38) to type BOOLEAN");
+        assertParsingError("cast(bigInteger1 as boolean)",
+                "Cast function cannot convert value of type DECIMAL(38, 38) to type BOOLEAN");
+
+        assertDataError("cast(string1 as boolean)", "VARCHAR value cannot be converted to BOOLEAN: 1");
+        assertRow("cast(stringFalse as boolean)", EXPR0, BOOLEAN, false);
+        assertDataError("cast(char1 as boolean)", "VARCHAR value cannot be converted to BOOLEAN: 1");
+
+        assertDataError("cast(object as boolean)", "Cannot convert OBJECT to BOOLEAN");
+    }
+
+    @Test
+    public void testTinyint() {
+        assertParsingError("cast(booleanTrue as tinyint)", "Cast function cannot convert value of type BOOLEAN to type TINYINT");
+
+        assertRow("cast(byte1 as tinyint)", EXPR0, TINYINT, (byte) 1);
+        assertRow("cast(byteMax as tinyint)", EXPR0, TINYINT, Byte.MAX_VALUE);
+        assertRow("cast(short1 as tinyint)", EXPR0, TINYINT, (byte) 1);
+        assertDataError("cast(shortMax as tinyint)", "Numeric overflow while converting SMALLINT to TINYINT: 32767");
+        assertRow("cast(int1 as tinyint)", EXPR0, TINYINT, (byte) 1);
+        assertDataError("cast(intMax as tinyint)", "Numeric overflow while converting INT to TINYINT: 2147483647");
+        assertRow("cast(long1 as tinyint)", EXPR0, TINYINT, (byte) 1);
+        assertDataError("cast(longMax as tinyint)", "Numeric overflow while converting BIGINT to TINYINT: 9223372036854775807");
+
+        assertRow("cast(float1 as tinyint)", EXPR0, TINYINT, (byte) 1);
+        assertDataError("cast(floatMax as tinyint)", "Numeric overflow while converting REAL to TINYINT: 3.4028235E38");
+        assertRow("cast(double1 as tinyint)", EXPR0, TINYINT, (byte) 1);
+        assertDataError("cast(doubleMax as tinyint)",
+                "Numeric overflow while converting DOUBLE to TINYINT: 1.7976931348623157E308");
+
+        assertRow("cast(decimal1 as tinyint)", EXPR0, TINYINT, (byte) 1);
+        assertDataError("cast(decimalBig as tinyint)",
+                "Numeric overflow while converting DECIMAL to TINYINT: 92233720368547758070");
+        assertRow("cast(bigInteger1 as tinyint)", EXPR0, TINYINT, (byte) 1);
+        assertDataError("cast(bigIntegerBig as tinyint)",
+                "Numeric overflow while converting DECIMAL to TINYINT: 92233720368547758070");
+
+        assertRow("cast(string1 as tinyint)", EXPR0, TINYINT, (byte) 1);
+        assertDataError("cast(stringBig as tinyint)", "Cannot convert VARCHAR to TINYINT: 92233720368547758070");
+        assertDataError("cast(stringFoo as tinyint)", "Cannot convert VARCHAR to TINYINT: foo");
+        assertRow("cast(char1 as tinyint)", EXPR0, TINYINT, (byte) 1);
+        assertDataError("cast(charF as tinyint)", "Cannot convert VARCHAR to TINYINT: f");
+
+        assertDataError("cast(object as tinyint)", "Cannot convert OBJECT to TINYINT");
```

Comment:
With every other source type we include the value that failed to convert in the error message, we should do that same with OBJECT.

Reply:
Due to [this comment](https://github.com/hazelcast/hazelcast/pull/17183#discussion_r464227192), let's just include the value class in the message.

Reply:
fixed in 9082f7d2b8af4d33049b70620c22f03e03150206